src/db/car/any.rs

```diff
@@ -1,25 +1,43 @@
 // Copyright 2019-2023 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
+
+//! There are three different CAR formats: `.car`, `.car.zst` and
+//! `.forest.car.zst`. [`AnyCar`] identifies the format by inspecting the CAR
+//! header and the first key-value block, and picks the appropriate block store
+//! (either [`super::ForestCar`] or [`super::PlainCar`]).
+//!
+//! CARv2 is not supported yet.
+
+use super::{CacheKey, ZstdFrameCache};
+use crate::blocks::Tipset;
 use cid::Cid;
 use fvm_ipld_blockstore::Blockstore;
+use parking_lot::Mutex;
 use std::io::{Cursor, Error, ErrorKind, Read, Result, Seek};
+use std::path::PathBuf;
+use std::sync::Arc;
 
 pub enum AnyCar<ReaderT> {
     Plain(super::PlainCar<ReaderT>),
     Forest(super::ForestCar<ReaderT>),
     Memory(super::PlainCar<Cursor<Vec<u8>>>),
 }
 
-impl<ReaderT: Read + Seek> AnyCar<ReaderT> {
-    pub fn new(mk_reader: impl Fn() -> Result<ReaderT> + Clone + 'static) -> Result<Self> {
-        if let Ok(forest_car) = super::ForestCar::new(mk_reader.clone()) {
+impl<ReaderT: super::CarReader> AnyCar<ReaderT> {
+    /// Open an archive. May be formatted as `.car`, `.car.zst` or
+    /// `.forest.car.zst`. This call may block for an indeterminate amount of
+    /// time while data is decoded and indexed.
+    pub fn new(mk_reader: impl super::forest::ReaderGen<ReaderT>) -> Result<Self> {
+        let plain_reader = mk_reader();
+        let zstd_reader = mk_reader();
```

Comment:
`zstd_reader` does not seem to be used in plain car code branch, how about inlining these

Reply:
Cannot be moved. `mk_reader` is being consumed one line below.